app_dart/lib/src/request_handlers/scheduler/vacuum_stale_tasks.dart

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:cocoon_service/cocoon_service.dart';
+import 'package:github/github.dart' as gh;
+import 'package:meta/meta.dart';
+
+import '../../model/appengine/task.dart';
+import '../../service/datastore.dart';
+import '../../service/logging.dart';
+
+/// Vacuum stale tasks.
+///
+/// Occassionally, a build never gets processed by LUCI. To prevent tasks
+/// being stuck as "In Progress," this will return tasks to "New" if they have
+/// no updates after 3 hours.
+@immutable
+class VacuumStaleTasks extends RequestHandler<Body> {
+  const VacuumStaleTasks({
+    required super.config,
+    @visibleForTesting this.datastoreProvider = DatastoreService.defaultProvider,
+    @visibleForTesting this.nowValue,
+  });
+
+  final DatastoreServiceProvider datastoreProvider;
+
+  /// For testing, can be used to inject a deterministic time.
+  final DateTime? nowValue;
+
+  static const int kCommitLimit = 25;
```

Comment:
It looks like `kCommitLimit` is not being used anywhere yet?

Reply:
Removed! This will instead rely on the datastore.dart limits

Reply:
nit: I think its hard to grok that `queryRecentTasks` has a particular limit but otherwise LGTM!

Reply:
From here, yes. However, we want all datastore logic contained there so it's easier make changes. It's difficult to trace datastore issues when they're scattered throughout the codebase.